aiohttp/client.py

```diff
@@ -939,6 +939,17 @@ async def _ws_connect(
             assert conn is not None
             conn_proto = conn.protocol
             assert conn_proto is not None
+
+            # For WS connection the read_timeout must be either ws_timeout.ws_receive or greater
+            # None == no timeout, i.e. infinite timeout, so None is the max timeout possible
+            if ws_timeout.ws_receive is None:
+                # Reset regardless
+                conn_proto.read_timeout = None
+            elif conn_proto.read_timeout is not None:
+                conn_proto.read_timeout = max(
+                    ws_timeout.ws_receive, conn_proto.read_timeout
+                )
+
```

Comment:
Maybe I'm just tired and not understanding, but can't we just pass the timeout above at:
https://github.com/aio-libs/aiohttp/blob/28f126cb673c258d8369a4fc2bcdf0ea4eef7a22/aiohttp/client.py#L851-L863

Reply:
You can't do that. the self.request is an HTTP request to upgrade to WS(S). That has to be done with the currently requested timeout settings for the HTTP connection. Once upgraded, the connection becomes WSS with WS-specific timeout settings.

That's the core of the bug and the crux of the issue. If you modify the connection here, you run the HTTP request to upgrade with (potentially infinite receive timeout) WS timeout settings.

Reply:
WS is special that way that HTTP becomes WS. Nothing else in the library behaves that way, that's why it was missed - the connection morphs the semantics from HTTP behavior to WS behavior.